packages/protocol/contracts/L1/v1/V1Verifying.sol

```diff
@@ -0,0 +1,152 @@
+// SPDX-License-Identifier: MIT
+//
+// ╭━━━━╮╱╱╭╮╱╱╱╱╱╭╮╱╱╱╱╱╭╮
+// ┃╭╮╭╮┃╱╱┃┃╱╱╱╱╱┃┃╱╱╱╱╱┃┃
+// ╰╯┃┃┣┻━┳┫┃╭┳━━╮┃┃╱╱╭━━┫╰━┳━━╮
+// ╱╱┃┃┃╭╮┣┫╰╯┫╭╮┃┃┃╱╭┫╭╮┃╭╮┃━━┫
+// ╱╱┃┃┃╭╮┃┃╭╮┫╰╯┃┃╰━╯┃╭╮┃╰╯┣━━┃
+// ╱╱╰╯╰╯╰┻┻╯╰┻━━╯╰━━━┻╯╰┻━━┻━━╯
+pragma solidity ^0.8.9;
+
+import "../../common/AddressResolver.sol";
+import "../TkoToken.sol";
+import "./V1Utils.sol";
+
+/// @author dantaik <dan@taiko.xyz>
+library V1Verifying {
+    using SafeCastUpgradeable for uint256;
+    event BlockVerified(uint256 indexed id, bytes32 blockHash);
+
+    event HeaderSynced(
+        uint256 indexed height,
+        uint256 indexed srcHeight,
+        bytes32 srcHash
+    );
+
+    function init(
+        LibData.State storage s,
+        bytes32 _genesisBlockHash,
+        uint256 _feeBase
+    ) public {
+        require(_feeBase > 0, "L1:feeBase");
+
+        s.genesisHeight = uint64(block.number);
+        s.genesisTimestamp = uint64(block.timestamp);
+        s.feeBase = _feeBase;
+        s.nextBlockId = 1;
+        s.lastProposedAt = uint64(block.timestamp);
+        s.l2Hashes[0] = _genesisBlockHash;
+
+        emit BlockVerified(0, _genesisBlockHash);
+        emit HeaderSynced(block.number, 0, _genesisBlockHash);
+    }
+
+    function verifyBlocks(
+        LibData.State storage s,
+        AddressResolver resolver,
+        uint256 maxBlocks,
+        bool checkHalt
+    ) public {
+        bool halted = V1Utils.isHalted(s);
+        if (checkHalt) {
+            require(!halted, "L1:halted");
+        } else if (halted) {
+            // skip finalizing blocks
+            return;
+        }
+
+        uint64 latestL2Height = s.latestVerifiedHeight;
+        bytes32 latestL2Hash = s.l2Hashes[latestL2Height];
+        uint64 processed = 0;
+        TkoToken tkoToken;
+
+        for (
+            uint256 i = s.latestVerifiedId + 1;
+            i < s.nextBlockId && processed <= maxBlocks;
+            i++
+        ) {
+            LibData.ForkChoice storage fc = s.forkChoices[i][latestL2Hash];
+
+            // Uncle proof can not take more than 2x time the first proof did.
+            if (
+                fc.blockHash == 0 ||
+                block.timestamp <= V1Utils.uncleProofDeadline(s, fc)
+            ) {
+                break;
+            } else {
+                if (fc.blockHash != LibConstants.K_BLOCK_DEADEND_HASH) {
+                    latestL2Height += 1;
+                    latestL2Hash = fc.blockHash;
+                }
+
+                if (LibConstants.K_TOKENOMICS_ENABLED) {
```

Comment:
this flag is only here for testing purpose, if we want to test core logics other than the tokenomics, in the test deployment, we can give TOKENOMICS_ENABLED False, to skip this part of the code. In production, this will always be True.